x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/Ccr.java

```diff
@@ -259,6 +281,119 @@ public Ccr(final Settings settings) {
     }
 
+    @Override
+    public Map<String, Repository.Factory> getRepositories(Environment env, NamedXContentRegistry namedXContentRegistry) {
+        Repository.Factory repositoryFactory = (metadata) -> new CcrRepository(metadata, settings);
+        return Collections.singletonMap(CcrRepository.TYPE, repositoryFactory);
+    }
+
     protected XPackLicenseState getLicenseState() { return XPackPlugin.getSharedLicenseState(); }
 
+    private static class CCRRepositoryManager extends RemoteClusterAware implements LocalNodeMasterListener {
+
+        private static final Logger LOGGER = LogManager.getLogger(CCRRepositoryManager.class);
+        private static final String SOURCE = "refreshing " + CcrRepository.TYPE + " repositories";
+
+        private final ClusterService clusterService;
+        private final Set<String> clusters = ConcurrentCollections.newConcurrentSet();
+        private volatile boolean isMasterNode = false;
+
+        private CCRRepositoryManager(Settings settings, ClusterService clusterService) {
+            super(settings);
+            this.clusterService = clusterService;
+            clusters.addAll(buildRemoteClustersDynamicConfig(settings).keySet());
+            clusterService.addLocalNodeMasterListener(this);
+            listenForUpdates(clusterService.getClusterSettings());
+        }
+
+        @Override
+        protected Set<String> getRemoteClusterNames() {
+            return clusters;
+        }
+
+        @Override
+        protected void updateRemoteCluster(String clusterAlias, List<String> addresses, String proxyAddress) {
+            if (addresses.isEmpty()) {
+                if (clusters.remove(clusterAlias) && isMasterNode) {
+                    refreshCCRRepositories();
+                }
+            } else {
+                if (clusters.add(clusterAlias) && isMasterNode) {
+                    refreshCCRRepositories();
+                }
+            }
+        }
+
+        @Override
+        public void onMaster() {
+            this.isMasterNode = true;
+            refreshCCRRepositories();
+        }
+
+        @Override
+        public void offMaster() {
+            this.isMasterNode = false;
+        }
+
+        @Override
+        public String executorName() {
+            return ThreadPool.Names.SAME;
+        }
+
+        private void refreshCCRRepositories() {
+            clusterService.submitStateUpdateTask(SOURCE, new ClusterStateUpdateTask() {
+
+                @Override
+                public ClusterState execute(ClusterState currentState) throws Exception {
+                    MetaData metaData = currentState.metaData();
+                    RepositoriesMetaData repositories = metaData.custom(RepositoriesMetaData.TYPE);
+
+                    if (repositories == null) {
+                        List<RepositoryMetaData> repositoriesMetaData = new ArrayList<>(clusters.size());
+                        for (String cluster : clusters) {
```

Comment:
instead of relying on a cached version, maybe parse the cluster state and get the current list of remote clusters? we can maybe add static methods in the RemoteClusterAware class to parse the settings into a whatever we need